lib/std/array_hash_map.zig

```diff
@@ -320,40 +358,54 @@ pub fn ArrayHashMap(
 pub fn ArrayHashMapUnmanaged(
     comptime K: type,
     comptime V: type,
-    comptime hash: fn (key: K) u32,
-    comptime eql: fn (a: K, b: K) bool,
+    comptime Context: type,
     comptime store_hash: bool,
 ) type {
+    comptime std.hash_map.verifyContext(Context, K, K, u32);
     return struct {
         /// It is permitted to access this field directly.
-        entries: std.ArrayListUnmanaged(Entry) = .{},
+        entries: DataList = .{},
 
         /// When entries length is less than `linear_scan_max`, this remains `null`.
         /// Once entries length grows big enough, this field is allocated. There is
         /// an IndexHeader followed by an array of Index(I) structs, where I is defined
         /// by how many total indexes there are.
         index_header: ?*IndexHeader = null,
 
-        /// Modifying the key is illegal behavior.
+        /// Modifying the key is allowed only if it does not change the hash.
         /// Modifying the value is allowed.
         /// Entry pointers become invalid whenever this ArrayHashMap is modified,
         /// unless `ensureCapacity` was previously used.
         pub const Entry = struct {
-            /// This field is `void` if `store_hash` is `false`.
+            key_ptr: *K,
+            value_ptr: *V,
+        };
+
+        pub const KV = struct {
+            key: K,
+            value: V,
+        };
+
+        pub const Data = struct {
             hash: Hash,
             key: K,
             value: V,
         };
 
+        pub const DataList = std.MultiArrayList(Data);
+
         pub const Hash = if (store_hash) u32 else void;
 
         pub const GetOrPutResult = struct {
-            entry: *Entry,
+            key_ptr: *K,
```

Comment:
Should this be `*const K`? You don't want people modifying the key...

Reply:
Modifying the key is a valid use case, and sometimes (in the case of getOrPut*Adapted) required.  But even aside from that, you might want to do a `put` with a temporary key, and then replace the key with a newly allocated version if the map took ownership of it.  `BufMap.put` does this, for example.